backends/zypp/pk-backend-zypp.cpp

```diff
@@ -3297,6 +3297,15 @@ backend_update_packages_thread (PkBackendJob *job, GVariant *params, gpointer us
 	ResPool pool = zypp_build_pool (zypp, TRUE);
 	PkRestartEnum restart = PK_RESTART_ENUM_NONE;
 
+	for (ResPool::const_iterator it = pool.begin (); it != pool.end (); ++it) {
+		if ( it->isKind<Product>() ) {
```

Comment:
Basically right. As you have some 1000 packages and just a few products you should `break;`  after having found `setUpgradeMode(TRUE)`. You may also use the pools internal filtering to iterate over the products only:
```
  for (ResPool::byKind_iterator it = pool.byKindBegin<Product>(); it != pool.byKindEnd<Product>(); ++it) {
    static const ...
    if ( (*it)->asKind<Product>()...
```
Or range based
```
  for ( const PoolItem & pi : pool.byKind<Product>() ) {
    static const ...
    if ( pi->asKind<Product>()...
```